### ISSUE TITLE
"Dangling" messages on stream

### ISSUE THREAD

**Issue by Kaarel** [nats-io/nats-server#3612]
nats-server: synadia/nats-server:nightly-20221107 (Git: [4e8d1ae])
nats: 0.0.34
3 node cluster, AWS EC2 Ubuntu

We have a interest based stream with 2 consumers, push subscribe, auto ack. We see no issues in our consuming services logs. But a lot of messages in the stream seemingly not consumed. Though vast majority of messages do get removed from the stream. Stream collects messages that are published using NATS core publish by a number of different services on different hosts. When issuing `nats stream view logs` it seems to clear most messages in the stream. But new ones continue to accumulate immediately.

1. `nats stream info logs; nats consumer info logs c1; nats consumer info logs c2;`

```
Information for Stream logs created 2022-07-13 20:07:45

             Subjects: logs.>
             Replicas: 3
              Storage: File

Options:

            Retention: Interest
     Acknowledgements: true
       Discard Policy: Old
     Duplicate Window: 2m0s
    Allows Msg Delete: true
         Allows Purge: true
       Allows Rollups: false

Limits:

     Maximum Messages: unlimited
  Maximum Per Subject: unlimited
        Maximum Bytes: 1.0 GiB
          Maximum Age: 5d0h0m0s
 Maximum Message Size: unlimited
    Maximum Consumers: unlimited


Cluster Information:

                 Name: nats-cluster
               Leader: n1
              Replica: n3, current, seen 0.07s ago
              Replica: n2, current, seen 0.07s ago

State:

             Messages: 23,229
                Bytes: 20 MiB
             FirstSeq: 134,390,673 @ 2022-11-07T21:52:23 UTC
              LastSeq: 134,849,300 @ 2022-11-08T08:23:30 UTC
     Deleted Messages: 435,399
     Active Consumers: 2
   Number of Subjects: 20

Information for Consumer logs > c1 created 2022-11-04T14:04:16Z

Configuration:

        Durable Name: c1
    Delivery Subject: c1.logs
      Deliver Policy: All
 Deliver Queue Group: c1
          Ack Policy: Explicit
            Ack Wait: 30s
       Replay Policy: Instant
     Max Ack Pending: 1,000
        Flow Control: false

Cluster Information:

                Name: nats-cluster
              Leader: n1
             Replica: n3, current, seen 0.17s ago
             Replica: n2, current, seen 0.17s ago

State:

   Last Delivered Message: Consumer sequence: 134,827,928 Stream sequence: 134,849,303 Last delivery: 0.17s ago
     Acknowledgment floor: Consumer sequence: 134,827,928 Stream sequence: 134,849,303 Last Ack: 0.17s ago
         Outstanding Acks: 0 out of maximum 1,000
     Redelivered Messages: 0
     Unprocessed Messages: 0
          Active Interest: Active using Queue Group c1
Information for Consumer logs > c2 created 2022-11-04T15:36:04Z

Configuration:

        Durable Name: c2
    Delivery Subject: c2.logs
      Deliver Policy: All
 Deliver Queue Group: c2
          Ack Policy: Explicit
            Ack Wait: 30s
       Replay Policy: Instant
     Max Ack Pending: 1,000
        Flow Control: false

Cluster Information:

                Name: nats-cluster
              Leader: n2
             Replica: n3, current, seen 0.50s ago
             Replica: n1, current, seen 0.50s ago

State:

   Last Delivered Message: Consumer sequence: 134,835,616 Stream sequence: 134,849,303 Last delivery: 0.50s ago
     Acknowledgment floor: Consumer sequence: 134,835,616 Stream sequence: 134,849,303 Last Ack: 0.50s ago
         Outstanding Acks: 0 out of maximum 1,000
     Redelivered Messages: 0
     Unprocessed Messages: 0
          Active Interest: Active using Queue Group c2
```

2. `nats stream view logs`

```
...
[messages removed for brevity]
...

? Next Page? No
```

3. `nats stream info logs; nats consumer info logs c1; nats consumer info logs c2;`

```
Information for Stream logs created 2022-07-13 20:07:45

             Subjects: logs.>
             Replicas: 3
              Storage: File

Options:

            Retention: Interest
     Acknowledgements: true
       Discard Policy: Old
     Duplicate Window: 2m0s
    Allows Msg Delete: true
         Allows Purge: true
       Allows Rollups: false

Limits:

     Maximum Messages: unlimited
  Maximum Per Subject: unlimited
        Maximum Bytes: 1.0 GiB
          Maximum Age: 5d0h0m0s
 Maximum Message Size: unlimited
    Maximum Consumers: unlimited


Cluster Information:

                 Name: nats-cluster
               Leader: n1
              Replica: n3, current, seen 1.30s ago
              Replica: n2, current, seen 1.31s ago

State:

             Messages: 35
                Bytes: 33 KiB
             FirstSeq: 134,848,747 @ 2022-11-08T08:22:17 UTC
              LastSeq: 134,849,345 @ 2022-11-08T08:23:42 UTC
     Deleted Messages: 564
     Active Consumers: 2
   Number of Subjects: 5

Information for Consumer logs > c1 created 2022-11-04T14:04:16Z

Configuration:

        Durable Name: c1
    Delivery Subject: c1.logs
      Deliver Policy: All
 Deliver Queue Group: c1
          Ack Policy: Explicit
            Ack Wait: 30s
       Replay Policy: Instant
     Max Ack Pending: 1,000
        Flow Control: false

Cluster Information:

                Name: nats-cluster
              Leader: n1
             Replica: n3, current, seen 0.00s ago
             Replica: n2, current, seen 0.00s ago

State:

   Last Delivered Message: Consumer sequence: 134,827,973 Stream sequence: 134,849,348 Last delivery: 0.04s ago
     Acknowledgment floor: Consumer sequence: 134,827,973 Stream sequence: 134,849,348 Last Ack: 0.02s ago
         Outstanding Acks: 0 out of maximum 1,000
     Redelivered Messages: 0
     Unprocessed Messages: 0
          Active Interest: Active using Queue Group c1
Information for Consumer logs > c2 created 2022-11-04T15:36:04Z

Configuration:

        Durable Name: c2
    Delivery Subject: c2.logs
      Deliver Policy: All
 Deliver Queue Group: c2
          Ack Policy: Explicit
            Ack Wait: 30s
       Replay Policy: Instant
     Max Ack Pending: 1,000
        Flow Control: false

Cluster Information:

                Name: nats-cluster
              Leader: n2
             Replica: n3, current, seen 0.27s ago
             Replica: n1, current, seen 0.27s ago

State:

   Last Delivered Message: Consumer sequence: 134,835,662 Stream sequence: 134,849,349 Last delivery: 0.27s ago
     Acknowledgment floor: Consumer sequence: 134,835,662 Stream sequence: 134,849,349 Last Ack: 0.27s ago
         Outstanding Acks: 0 out of maximum 1,000
     Redelivered Messages: 0
     Unprocessed Messages: 0
          Active Interest: Active using Queue Group c2
```

4. `nats stream purge logs; nats stream info logs; nats consumer info logs c1; nats consumer info logs c2;`

```
? Really purge Stream logs Yes
Information for Stream logs created 2022-07-13 20:07:45

             Subjects: logs.>
             Replicas: 3
              Storage: File

Options:

            Retention: Interest
     Acknowledgements: true
       Discard Policy: Old
     Duplicate Window: 2m0s
    Allows Msg Delete: true
         Allows Purge: true
       Allows Rollups: false

Limits:

     Maximum Messages: unlimited
  Maximum Per Subject: unlimited
        Maximum Bytes: 1.0 GiB
          Maximum Age: 5d0h0m0s
 Maximum Message Size: unlimited
    Maximum Consumers: unlimited


Cluster Information:

                 Name: nats-cluster
               Leader: n1
              Replica: n3, current, seen 0.00s ago
              Replica: n2, outdated, seen 0.13s ago, 1 operation behind

State:

             Messages: 0
                Bytes: 0 B
             FirstSeq: 134,857,694 @ 0001-01-01T00:00:00 UTC
              LastSeq: 134,857,693 @ 2022-11-08T08:39:56 UTC
     Active Consumers: 2
Information for Stream logs created 2022-07-13 20:07:45

             Subjects: logs.>
             Replicas: 3
              Storage: File

Options:

            Retention: Interest
     Acknowledgements: true
       Discard Policy: Old
     Duplicate Window: 2m0s
    Allows Msg Delete: true
         Allows Purge: true
       Allows Rollups: false

Limits:

     Maximum Messages: unlimited
  Maximum Per Subject: unlimited
        Maximum Bytes: 1.0 GiB
          Maximum Age: 5d0h0m0s
 Maximum Message Size: unlimited
    Maximum Consumers: unlimited


Cluster Information:

                 Name: nats-cluster
               Leader: n1
              Replica: n3, current, seen 0.27s ago
              Replica: n2, current, seen 0.27s ago

State:

             Messages: 0
                Bytes: 0 B
             FirstSeq: 134,857,695
              LastSeq: 134,857,694 @ 2022-11-08T08:39:56 UTC
     Active Consumers: 2

Information for Consumer logs > c1 created 2022-11-04T14:04:16Z

Configuration:

        Durable Name: c1
    Delivery Subject: c1.logs
      Deliver Policy: All
 Deliver Queue Group: c1
          Ack Policy: Explicit
            Ack Wait: 30s
       Replay Policy: Instant
     Max Ack Pending: 1,000
        Flow Control: false

Cluster Information:

                Name: nats-cluster
              Leader: n1
             Replica: n3, current, seen 0.09s ago
             Replica: n2, current, seen 0.09s ago

State:

   Last Delivered Message: Consumer sequence: 134,836,321 Stream sequence: 134,857,696 Last delivery: 0.10s ago
     Acknowledgment floor: Consumer sequence: 134,836,321 Stream sequence: 134,857,696 Last Ack: 0.10s ago
         Outstanding Acks: 0 out of maximum 1,000
     Redelivered Messages: 0
     Unprocessed Messages: 0
          Active Interest: Active using Queue Group c1
Information for Consumer logs > c2 created 2022-11-04T15:36:04Z

Configuration:

        Durable Name: c2
    Delivery Subject: c2.logs
      Deliver Policy: All
 Deliver Queue Group: c2
          Ack Policy: Explicit
            Ack Wait: 30s
       Replay Policy: Instant
     Max Ack Pending: 1,000
        Flow Control: false

Cluster Information:

                Name: nats-cluster
              Leader: n2
             Replica: n3, current, seen 0.41s ago
             Replica: n1, current, seen 0.40s ago

State:

   Last Delivered Message: Consumer sequence: 134,844,004 Stream sequence: 134,857,696 Last delivery: 0.41s ago
     Acknowledgment floor: Consumer sequence: 134,844,004 Stream sequence: 134,857,696 Last Ack: 0.41s ago
         Outstanding Acks: 0 out of maximum 1,000
     Redelivered Messages: 0
     Unprocessed Messages: 0
          Active Interest: Active using Queue Group c2
```

5. `nats stream info logs; nats consumer info logs c1; nats consumer info logs c2;`

```
Information for Stream logs created 2022-07-13 20:07:45

             Subjects: logs.>
             Replicas: 3
              Storage: File

Options:

            Retention: Interest
     Acknowledgements: true
       Discard Policy: Old
     Duplicate Window: 2m0s
    Allows Msg Delete: true
         Allows Purge: true
       Allows Rollups: false

Limits:

     Maximum Messages: unlimited
  Maximum Per Subject: unlimited
        Maximum Bytes: 1.0 GiB
          Maximum Age: 5d0h0m0s
 Maximum Message Size: unlimited
    Maximum Consumers: unlimited


Cluster Information:

                 Name: nats-cluster
               Leader: n1
              Replica: n3, current, seen 0.29s ago
              Replica: n2, current, seen 0.29s ago

State:

             Messages: 44
                Bytes: 40 KiB
             FirstSeq: 134,857,701 @ 2022-11-08T08:39:59 UTC
              LastSeq: 134,858,552 @ 2022-11-08T08:41:21 UTC
     Deleted Messages: 808
     Active Consumers: 2
   Number of Subjects: 6

Information for Consumer logs > c1 created 2022-11-04T14:04:16Z

Configuration:

        Durable Name: c1
    Delivery Subject: c1.logs
      Deliver Policy: All
 Deliver Queue Group: c1
          Ack Policy: Explicit
            Ack Wait: 30s
       Replay Policy: Instant
     Max Ack Pending: 1,000
        Flow Control: false

Cluster Information:

                Name: nats-cluster
              Leader: n1
             Replica: n3, current, seen 0.61s ago
             Replica: n2, current, seen 0.61s ago

State:

   Last Delivered Message: Consumer sequence: 134,837,177 Stream sequence: 134,858,552 Last delivery: 0.62s ago
     Acknowledgment floor: Consumer sequence: 134,837,177 Stream sequence: 134,858,552 Last Ack: 0.61s ago
         Outstanding Acks: 0 out of maximum 1,000
     Redelivered Messages: 0
     Unprocessed Messages: 0
          Active Interest: Active using Queue Group c1
Information for Consumer logs > c2 created 2022-11-04T15:36:04Z

Configuration:

        Durable Name: c2
    Delivery Subject: c2.logs
      Deliver Policy: All
 Deliver Queue Group: c2
          Ack Policy: Explicit
            Ack Wait: 30s
       Replay Policy: Instant
     Max Ack Pending: 1,000
        Flow Control: false

Cluster Information:

                Name: nats-cluster
              Leader: n2
             Replica: n3, current, seen 0.16s ago
             Replica: n1, current, seen 0.16s ago

State:

   Last Delivered Message: Consumer sequence: 134,844,859 Stream sequence: 134,858,554 Last delivery: 0.16s ago
     Acknowledgment floor: Consumer sequence: 134,844,859 Stream sequence: 134,858,554 Last Ack: 0.16s ago
         Outstanding Acks: 0 out of maximum 1,000
     Redelivered Messages: 0
     Unprocessed Messages: 0
          Active Interest: Active using Queue Group c2
```

From then on messages and deleted messages count on the stream continues to increase. Delete messages count increases faster than messages count.

**Comment by derekcollison** [comment]
So the acks and the ack floor look correct but the stream did not clean a message (some messages) below the ack floor. 

Messages published into the stream, are they NATS core messages or do the apps wait on the pubAck?

**Comment by derekcollison** [comment]
Also how fast are the messages ack'd in relation to them being published?

**Comment by Kaarel** [comment]
> Messages published into the stream, are they NATS core messages or do the apps wait on the pubAck?

These are all NATS core messages. Some services originate pre JetStream and some are older client libraries even today.

> Also how fast are the messages ack'd in relation to them being published?

I hope I understand the question, but this is how the messages flow:
1. NATS core publish (no pubAck)
2. Message gets picked up by the stream
3. Consuming service gets notified immediately (push subscribe)
4. Consuming service saves the message into a persistent storage and acks the message even in case of error.

More specific details on point 4. Consumer is a dotnet service. Subscribing to stream using method `PushSubscribeAsync` with `autoAck` parameter set to true. The message handler is wrapped in `try {} catch {}` so even if there were errors parsing the message, it should be acked.

**Comment by derekcollison** [comment]
What is the approximate ingest rate to the stream?

**Comment by derekcollison** [comment]
Also do the subscribers stay constant or do other consumers come and go from time to time?

**Comment by Kaarel** [comment]
> What is the approximate ingest rate to the stream?

Using the NATS JetStream Grafana Dashboard:

![pilt](https://user-images.githubusercontent.com/108327/201027499-67fbe6f8-bef2-4b7d-b605-f4e1eb2d342c.png)

And without the three spikes
![pilt](https://user-images.githubusercontent.com/108327/201027571-633549e5-1f76-4b62-a849-c225775cdb8c.png)

**Comment by Kaarel** [comment]
> Could you share the .NET client version and the actual `PushSubscribeAsync` call params?

NATS.Client 1.0.1

```
var options = PushSubscribeOptions.BindTo("logs", "c1");
var subscription = _connection.CreateJetStreamContext().PushSubscribeAsync(
    "logs.>",
    "c1",
    (sender, args) => { try { } catch (Exception) { } },
    true,
    options);
```

**Comment by Kaarel** [comment]
> Also do the subscribers stay constant or do other consumers come and go from time to time?

Subscribers are 2 services that run 24/7. Only change when redeploying.

**Comment by derekcollison** [comment]
Thanks for the feedback.

And you see dangling messages without doing a server or cluster restart and no new consumers..

**Comment by Kaarel** [comment]
> And you see dangling messages without doing a server or cluster restart and no new consumers..

Correct.

Btw when before `nats stream view` used to clear out messages, with the nightly build it always leaves a few dozen messages and only `nats stream purge` clears all messages. But the message count starts to grow again almost immediately. Though at a much much slower rate than the deleted messages count in `nats stream report`.

**Comment by derekcollison** [comment]
Ah so you frequently do `nats stream view`? That creates ephemeral consumers when you execute that, that may be the hint I needed.

**Comment by Kaarel** [comment]
> Ah so you frequently do `nats stream view`? That creates ephemeral consumers when you execute that, that may be the hint I needed.

Well frequently, as whenever the disk space taken by the stream feels unreasonably high as it shouldn't really be continuously increasing with our cnfguration. Before the nightly it used to be once a week or so. Now it's been more like once a day or every other day. I can see the ehemeral consumer momentarily in stream `obs` directory. But it disappears quickly.

**Comment by derekcollison** [comment]
ok so you really only do this in response to a high disk usage, meaning the dangling messages have already been triggered it seems..

Any other events that you could point to that could be the trigger?

**Comment by Kaarel** [comment]
> ok so you really only do this in response to a high disk usage, meaning the dangling messages have already been triggered it seems..

Correct. That is the reason to issue `nats stream view` in the first place.

> Any other events that you could point to that could be the trigger?

None that I can think of. Nothing stands out to me. That's the odd bit, especially with the nightly when after `nats steam purge` the "dangling" message count starts to grow pretty much immediately.

**Comment by derekcollison** [comment]
Did purges, ttl and stream views and still ok for me. Which just means I am missing something and not hitting the condition, not that it does not exist..

**Comment by Kaarel** [comment]
Thanks, I appreciate your time! Overnight since last `stream purge` we've got 59,709 messages on the stream (should be close to 0) and 1000362 deleted messages according to `stream report`.

Short of full repro project is there any diagnostics flags that we can switch on that could help diagnose the issue.

**Comment by derekcollison** [comment]
Can you share stream and consumer info again for me?

**Comment by derekcollison** [comment]
I see that the consumers are replicated as well but also using queue groups. How many apps are servicing each consumer?

**Comment by derekcollison** [comment]
Would it be possible to jump on a zoom call? DM me if that could work.

**Comment by Kaarel** [comment]
> Can you share stream and consumer info again for me?

`nats stream info logs; nats consumer info logs c1`

```
Information for Stream logs created 2022-07-13 20:07:45

             Subjects: logs.>
             Replicas: 3
              Storage: File

Options:

            Retention: Interest
     Acknowledgements: true
       Discard Policy: Old
     Duplicate Window: 2m0s
    Allows Msg Delete: true
         Allows Purge: true
       Allows Rollups: false

Limits:

     Maximum Messages: unlimited
  Maximum Per Subject: unlimited
        Maximum Bytes: 1.0 GiB
          Maximum Age: 5d0h0m0s
 Maximum Message Size: unlimited
    Maximum Consumers: unlimited


Cluster Information:

                 Name: nats-cluster
               Leader: n1
              Replica: n3, current, seen 0.01s ago
              Replica: n2, current, seen 0.01s ago

State:

             Messages: 10,274
                Bytes: 8.8 MiB
             FirstSeq: 137,832,029 @ 2022-11-11T10:11:23 UTC
              LastSeq: 138,101,013 @ 2022-11-11T14:46:27 UTC
     Deleted Messages: 258,711
     Active Consumers: 2
   Number of Subjects: 19

Information for Consumer logs > c1 created 2022-11-04T14:04:16Z

Configuration:

        Durable Name: c1
    Delivery Subject: c1.logs
      Deliver Policy: All
 Deliver Queue Group: c1
          Ack Policy: Explicit
            Ack Wait: 30s
       Replay Policy: Instant
     Max Ack Pending: 1,000
        Flow Control: false

Cluster Information:

                Name: nats-cluster
              Leader: n1
             Replica: n3, current, seen 0.34s ago
             Replica: n2, current, seen 0.34s ago

State:

   Last Delivered Message: Consumer sequence: 138,067,772 Stream sequence: 138,101,013 Last delivery: 0.35s ago
     Acknowledgment floor: Consumer sequence: 138,067,772 Stream sequence: 138,101,013 Last Ack: 0.35s ago
         Outstanding Acks: 0 out of maximum 1,000
     Redelivered Messages: 0
     Unprocessed Messages: 0
          Active Interest: Active using Queue Group c1
```

**Comment by Kaarel** [comment]
> I see that the consumers are replicated as well but also using queue groups. How many apps are servicing each consumer?

There is one app per consumer. And one instance of the app.

**Comment by Kaarel** [comment]
> Would it be possible to jump on a zoom call? DM me if that could work.

That would be good! In practice I'm travelling through different time zones atm. Will know more about my schedule in the new week.

**Comment by derekcollison** [comment]
Ok keep us posted. I'll keep trying things to reproduce.

**Comment by derekcollison** [comment]
Think I found it.

**Comment by Kaarel** [comment]
> Think I found it.

Wow that was fast! I'm still recovering from the 9h TZ diff :) but using nats nightly-20221115 (Git: [095b13c4]) for a day the dangling message issue is gone and as a bonus memory usage seems to be noticeably lower as well.

Awesome work!

**Comment by derekcollison** [comment]
Will be included in 2.9.7, which we will release this week.

**Comment by derekcollison** [comment]
The stream view will create a consumer and when that consumer clears the system feels that all acks have been received for the message.

So it sounds like there is still an edge condition where an ack is processed by the system but messages are not properly cleaned up.

**Comment by derekcollison** [comment]
Would be good to verify that when the condition happens, `nats consumer report` does show the ack floor for all attached consumers above the dangling messages, just to confirm.

**Comment by Kaarel** [comment]
`nats consumer report` before and after `nats stream view` clears the stream? Will do that next time.

**Comment by derekcollison** [comment]
`nats stream view` will clear it because it creates and then deletes a consumer from the beginning of the stream, so the logic will look at those messages when removing the new consumer and see no more consumers are interested in a given message or they have all ack'd the message.

**Comment by Kaarel** [comment]
It took about a day this time.

Summary of steps:
1. `nats stream report; nats stream info logs; nats consumer report logs; nats consumer info logs c1;`
2. `nats stream view logs`
3. `nats stream report; nats stream info logs; nats consumer report logs; nats consumer info logs c1;`

Step details:

1. `nats stream report; nats stream info logs; nats consumer report logs; nats consumer info logs c1;`
```
Obtaining Stream stats

╭─────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                            Stream Report                                        │
├───────────┬─────────┬───────────┬───────────┬──────────┬─────────┬──────┬─────────┬─────────────┤
│ Stream    │ Storage │ Placement │ Consumers │ Messages │ Bytes   │ Lost │ Deleted │ Replicas    │
├───────────┼─────────┼───────────┼───────────┼──────────┼─────────┼──────┼─────────┼─────────────┤
│ [...]     │ File    │           │ 1         │ 0        │ 0 B     │ 0    │ 0       │ n3, n1, n2* │
│ logs      │ File    │           │ 2         │ 3        │ 5.0 KiB │ 0    │ 868888  │ n3, n1, n2* │
╰───────────┴─────────┴───────────┴───────────┴──────────┴─────────┴──────┴─────────┴─────────────╯

Information for Stream logs created 2022-07-13 20:07:45

             Subjects: logs.>
             Replicas: 3
              Storage: File

Options:

            Retention: Interest
     Acknowledgements: true
       Discard Policy: Old
     Duplicate Window: 2m0s
    Allows Msg Delete: true
         Allows Purge: true
       Allows Rollups: false

Limits:

     Maximum Messages: unlimited
  Maximum Per Subject: unlimited
        Maximum Bytes: 1.0 GiB
          Maximum Age: 5d0h0m0s
 Maximum Message Size: unlimited
    Maximum Consumers: unlimited


Cluster Information:

                 Name: nats-cluster
               Leader: n2
              Replica: n3, current, seen 0.00s ago
              Replica: n1, current, seen 0.00s ago

State:

             Messages: 3
                Bytes: 5.0 KiB
             FirstSeq: 169,039,453 @ 2022-12-08T16:52:45 UTC
              LastSeq: 169,908,347 @ 2022-12-09T11:39:58 UTC
     Deleted Messages: 868,892
     Active Consumers: 2
   Number of Subjects: 2

╭──────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                 Consumer report for logs with 2 consumers                                    │
├─────────┬──────┬────────────┬──────────┬─────────────┬─────────────┬─────────────┬─────────────┬─────────────┤
│ Consumer│ Mode │ Ack Policy │ Ack Wait │ Ack Pending │ Redelivered │ Unprocessed │ Ack Floor   │ Cluster     │
├─────────┼──────┼────────────┼──────────┼─────────────┼─────────────┼─────────────┼─────────────┼─────────────┤
│ c2      │ Push │ Explicit   │ 30.00s   │ 0           │ 0           │ 0           │ 169,908,347 │ n3, n1*, n2 │
│ c1      │ Push │ Explicit   │ 30.00s   │ 0           │ 0           │ 0           │ 169,908,347 │ n3, n1, n2* │
╰─────────┴──────┴────────────┴──────────┴─────────────┴─────────────┴─────────────┴─────────────┴─────────────╯

Information for Consumer logs > c1 created 2022-11-15T06:21:30Z

Configuration:

        Durable Name: c1
    Delivery Subject: c1.logs
      Deliver Policy: All
 Deliver Queue Group: c1
          Ack Policy: Explicit
            Ack Wait: 30s
       Replay Policy: Instant
     Max Ack Pending: 1,000
        Flow Control: false

Cluster Information:

                Name: nats-cluster
              Leader: n2
             Replica: n3, current, seen 0.77s ago
             Replica: n1, current, seen 0.77s ago

State:

   Last Delivered Message: Consumer sequence: 169,837,846 Stream sequence: 169,908,347 Last delivery: 0.78s ago
     Acknowledgment floor: Consumer sequence: 169,837,846 Stream sequence: 169,908,347 Last Ack: 0.78s ago
         Outstanding Acks: 0 out of maximum 1,000
     Redelivered Messages: 0
     Unprocessed Messages: 0
          Active Interest: Active using Queue Group c1
```


2. `nats stream view logs`
```
[169039453] [...] Received: 2022-12-08T16:52:45Z

[169735262] [...] Received: 2022-12-09T07:16:21Z

[169844791] [...] Received: 2022-12-09T10:03:55Z

[169909458] [...] Received: 2022-12-09T11:40:48Z

[169909459] [...] Received: 2022-12-09T11:40:48Z

11:40:48 Reached apparent end of data
```

3. `nats stream report; nats stream info logs; nats consumer report logs; nats consumer info logs c1;`
```
Obtaining Stream stats

╭─────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                            Stream Report                                        │
├───────────┬─────────┬───────────┬───────────┬──────────┬─────────┬──────┬─────────┬─────────────┤
│ Stream    │ Storage │ Placement │ Consumers │ Messages │ Bytes   │ Lost │ Deleted │ Replicas    │
├───────────┼─────────┼───────────┼───────────┼──────────┼─────────┼──────┼─────────┼─────────────┤
│ [...]     │ File    │           │ 1         │ 0        │ 0 B     │ 0    │ 0       │ n3, n1, n2* │
│ logs      │ File    │           │ 2         │ 0        │ 0 B     │ 0    │ 0       │ n3, n1, n2* │
╰───────────┴─────────┴───────────┴───────────┴──────────┴─────────┴──────┴─────────┴─────────────╯

Information for Stream logs created 2022-07-13 20:07:45

             Subjects: logs.>
             Replicas: 3
              Storage: File

Options:

            Retention: Interest
     Acknowledgements: true
       Discard Policy: Old
     Duplicate Window: 2m0s
    Allows Msg Delete: true
         Allows Purge: true
       Allows Rollups: false

Limits:

     Maximum Messages: unlimited
  Maximum Per Subject: unlimited
        Maximum Bytes: 1.0 GiB
          Maximum Age: 5d0h0m0s
 Maximum Message Size: unlimited
    Maximum Consumers: unlimited


Cluster Information:

                 Name: nats-cluster
               Leader: n2
              Replica: n3, current, seen 0.79s ago
              Replica: n1, current, seen 0.79s ago

State:

             Messages: 0
                Bytes: 0 B
             FirstSeq: 169,909,716
              LastSeq: 169,909,715 @ 2022-12-09T11:41:13 UTC
     Active Consumers: 2

╭──────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                 Consumer report for logs with 2 consumers                                    │
├─────────┬──────┬────────────┬──────────┬─────────────┬─────────────┬─────────────┬─────────────┬─────────────┤
│ Consumer│ Mode │ Ack Policy │ Ack Wait │ Ack Pending │ Redelivered │ Unprocessed │ Ack Floor   │ Cluster     │
├─────────┼──────┼────────────┼──────────┼─────────────┼─────────────┼─────────────┼─────────────┼─────────────┤
│ c2      │ Push │ Explicit   │ 30.00s   │ 0           │ 0           │ 0           │ 169,909,715 │ n3, n1*, n2 │
│ c1      │ Push │ Explicit   │ 30.00s   │ 0           │ 0           │ 0           │ 169,909,715 │ n3, n1, n2* │
╰─────────┴──────┴────────────┴──────────┴─────────────┴─────────────┴─────────────┴─────────────┴─────────────╯

Information for Consumer logs > c1 created 2022-11-15T06:21:30Z

Configuration:

        Durable Name: c1
    Delivery Subject: c1.logs
      Deliver Policy: All
 Deliver Queue Group: c1
          Ack Policy: Explicit
            Ack Wait: 30s
       Replay Policy: Instant
     Max Ack Pending: 1,000
        Flow Control: false

Cluster Information:

                Name: nats-cluster
              Leader: n2
             Replica: n3, current, seen 1.50s ago
             Replica: n1, current, seen 1.49s ago

State:

   Last Delivered Message: Consumer sequence: 169,839,214 Stream sequence: 169,909,715 Last delivery: 1.50s ago
     Acknowledgment floor: Consumer sequence: 169,839,214 Stream sequence: 169,909,715 Last Ack: 1.50s ago
         Outstanding Acks: 0 out of maximum 1,000
     Redelivered Messages: 0
     Unprocessed Messages: 0
          Active Interest: Active using Queue Group c1
```

Some visualisations:
![pilt](https://user-images.githubusercontent.com/108327/206714735-94a8fee2-dcfe-47da-b7d5-43ffcec15b15.png)

![pilt](https://user-images.githubusercontent.com/108327/206714778-1fb00363-cc28-4d8d-820b-1ba2d60b1880.png)

![pilt](https://user-images.githubusercontent.com/108327/206714812-0d66de49-3294-4436-a306-349462207423.png)

**Comment by Kaarel** [comment]
We've NATS running without explicitly configured log level which I think then runs at default level INFO? There is no leader related events in the log file. In fact in the below screenshots the messages start being retained again on the 10th and there are 0 events on that date in all 3 cluster nodes. So nothing interesting in the logs at INFO level.

There does seem to be some sort of visual correlation, if relevant at all, between `Stream message count` and `Pending messages` Grafana dashboards.

Occasionally stream message count can be seen decreasing. This should be caused by stream limits setting max age 5 days.

![pilt](https://user-images.githubusercontent.com/108327/208475460-684360fd-834b-4317-a9d5-fda9373859b4.png)

![pilt](https://user-images.githubusercontent.com/108327/208475501-a07ab845-d109-461f-87b2-74b9ab3d8ea8.png)

Happy to configure logging to help investigate.

**Comment by derekcollison** [comment]
What server version are you running again?

**Comment by Kaarel** [comment]
> 

2.9.8 at the moment.

**Comment by derekcollison** [comment]
Strongly suggest being current on patch release. 2.9.10 will go out later today or tomorrow but you should be on 2.9.9 at least.

**Comment by Kaarel** [comment]
Upgraded to 2.9.10 today.

An interesting observation. Even though the issue is most apparent with the one actively used stream, it occasionally happens in other streams too that see very little traffic in comparison. For example just after upgrading to 2.9.10 Grafana dashboard showed 1 message on another stream. That stream has a similar configuration - messages are removed after consumption. What was unusual this time, and maybe a hint, is that `nats stream report` didn't show any messages on that stream. Only the dashboard, which I believe resolves via prometheus-nats-exporter to the monitoring endpoint `localhost:8222`. When issuing `nats stream view` on that stream I got no messages just a `11:53:40 Reached apparent end of data` on the console. After that the Grafana dashboard showed 0 messages again.

Will continue monitoring the system on 2.9.10.

**Comment by derekcollison** [comment]
Thanks for the feedback @Kaarel

**Comment by Kaarel** [comment]
New year with good news, hopefully :) NATS 2.9.10 has been running for 11 days now and no dangling messages so far! The overall profile has been more stable as well. Some visualisations from one of the cluster nodes below. Noteworthy I guess is the long lasting spikes in network traffic before the upgrade. This only on one cluster node, the others show a stable network traffic. Can't think of anything on our architecture that would have changed the same time we upgraded NATS so...

In any case good to see things stable and predictable!

![pilt](https://user-images.githubusercontent.com/108327/210209300-94377d6b-c9f5-4f9e-b1ad-4d45118cd205.png)

**Comment by Kaarel** [comment]
Think I'm happy to close this. The 2.9.10 cluster has been running for over a month without issues.

We just upgraded from 2.9.10 to 2.9.14 and ran into an unrelated issue described here #3858 

Interestingly this upgrade was followed by a short spike of these "dangling" messages again on the Grafana dashboard. No messages in `stream report`. And as before `nats stream view` cleared the issue.

Basically I followed every nats node upgrade with `nats stream view` and sans #3858 is looking stable.
